### DISCUSSION TITLE
Order of middlewareArguments

### DISCUSSION THREAD

**Discussion by lvd1999** [floating-ui/floating-ui#1965] [Q&A]
Hi

I have a middleArguments object to check overflow by using detectOverflow function on my tooltip. However, the tooltip seems that it doesn't touch shift() or flip(). Perhaps the order of middleware is wrong?

My computPosition: 
```javascript
computePosition(component, tooltip, {
      placement: this.position,
      middleware: [
        middleware,  // if placed here, arrow placing is always at the end of reference element (look wrong)
        offset(10),
        flip(),
        shift({ padding: 5 }),
        arrow({ element: tooltipArrow }),
       // middleware, // if placed here, better results on arrow if done at end, but shift() and flip() not touched
      ],
    }
      }

      const { x, y, placement, middlewareData } = data;
      Object.assign(tooltip.style, {
        left: `${x}px`,
        top: `${y}px`,
      });

      const { x: arrowX, y: arrowY } = middlewareData.arrow;
      const mainStaticSide = {
        top: 'bottom',
        right: 'left',
        bottom: 'top',
        left: 'right',
      }[placement.split('-')[0]];

      const crossStaticSide = {
        'right-start': 'top',
        'left-start': 'top',
        'right-end': 'bottom',
        'left-end': 'bottom',
        'top-end': 'right',
        'bottom-end': 'right',
      }[placement];

      const unsetSides = {
        top: '',
        left: '',
        right: '',
        bottom: '',
      };

      // is non-center aligned
      if (placement.split('-')[1]) {
        Object.assign(tooltipArrow.style, {
          ...unsetSides,
          [crossStaticSide]: `5px`,
          [mainStaticSide]: '-4px',
        });
      } else {
        Object.assign(tooltipArrow.style, {
          ...unsetSides,
          left: arrowX != null ? `${arrowX}px` : '',
          top: arrowY != null ? `${arrowY}px` : '',
          [mainStaticSide]: '-4px',
        });
      }
    });
  }
```

**Comment by Grandizer** [comment]
Did you ever figure this out, this is Exactly what I am needing. The documentation is rather lacking.

**Reply by atomiks** [reply]
Can you provide some code of what you're trying to do? What area is the docs lacking in?

**Reply by Grandizer** [reply]
Here is my code. I am trying to implement the `detectOverflow`:

    export function setupPopovers() {
      const triggers = document.querySelectorAll("[id^='my-trigger-']");
      const body = document.getElementById("the-body");

      triggers.forEach((triggerElement) => {
        const trigsPopId = triggerElement.getAttribute("popup-id");
        const arrowId = triggerElement.getAttribute("arrow-id");

        if (trigsPopId) {
          const popoverElement = document.getElementById(trigsPopId);
          const arrowElement = document.getElementById(arrowId);

          if (popoverElement) {
            const middlewareDO = {
              name: 'middleware',
              async fn(state) {
                const overflow = await detectOverflow(state, {
                  boundary: body,
                });
                console.log('should flip:', overflow.top >= 0); // This is the boolean I want
                return {}; // have tried returning overflow.top >= 0 but nothing shows
              },
            };

            function updatePosition(trigger, popover) {
              computePosition(trigger, popover, {
                placement: triggerElement.getAttribute("my-arrow-position"),
                middleware: [
                  offset(8),
                  middlewareDO,
                  flip({ crossAxis: true }),
                  shift({ padding: 5 }),
                  arrow({ element: arrowElement }),
                ],
              })
              .then(({ x, y, placement, middlewareData }) => {
                Object.assign(popover.style, {
                  left: `${x}px`,
                  top: `${y}px`,
                });

                console.log('middleware:', middlewareData);

                const { x: arrowX, y: arrowY } = middlewareData.arrow;
                const staticSide = {
                  top: 'bottom',
                  right: 'left',
                  bottom: 'top',
                  left: 'right',
                }[placement.split('-')[0]];

                Object.assign(arrowElement.style, {
                  left: arrowX != null ? `${arrowX}px` : '',
                  top: arrowY != null ? `${arrowY}px` : '',
                  right: '',
                  bottom: '',
                  [staticSide]: '-4px',
                });
              });
            }

            triggerElement.addEventListener('mouseenter', () => {
              updatePosition(triggerElement, popoverElement);
              popoverElement.style.display = 'block';
            });
            triggerElement.addEventListener('mouseleave', () => {
              popoverElement.style.display = 'none';
            });
          }
        }
      });
    }


In a situation where it does not need to flip this is the log: 

    overflow: {top: -3, bottom: -422.046875, left: -273.42949999999996, right: -740.4295}
    should flip: false
    middleware: 
      arrow: {x: 166.5, centerOffset: 0}
      flip: {}
      middleware: {}
      offset: {x: 0, y: -8}
      shift: {x: 0, y: 0}

And in the situation where it should flip I see this in the log:

    overflow: {top: 22, bottom: -447.046875, left: -273.42949999999996, right: -740.4295}
    should flip: true
    middleware: 
      arrow: {x: 166.5, centerOffset: 0}
      flip: {}
      middleware: {}
      offset: {x: 0, y: -8}
      shift: {x: 0, y: 0}

So how can I get the value from the `detectOverflow` into the `middlewareData` or somewhere accessible?

**Reply by atomiks** [reply]
This is documented here: https://floating-ui.com/docs/middleware#data

```js
const middlewareDO = {
  name: "middleware",
  async fn(state) {
    const overflow = await detectOverflow(state, {
      boundary: body,
    });
    return {
      data: {
        shouldFlip: overflow.top >= 0,
      }
    };
  },
};
```

Now it's available in `middlewareData.middleware.shouldFlip`

**Reply by Grandizer** [reply]
Thanks for the update. I did get it to work, the `data:` element was the key part I did not have. Sadly, I am now having an issue of how to flip or change the position of the popper now, based on if that `shouldFlip` is true. Any suggestions @atomiks ?
